cli/crates/backend/src/api/deploy.rs

```diff
@@ -19,26 +21,118 @@ use tokio_util::codec::{BytesCodec, FramedRead};
 
 const ENTRY_BLACKLIST: [&str; 2] = ["node_modules", ".env"];
 
+enum ProjectDefinition {
+    ById {
+        project_id: String,
+        branch: Option<String>,
+    },
+    BySlugs {
+        account_slug: String,
+        project_slug: String,
+        branch: Option<String>,
+    },
+}
+
 /// # Errors
 ///
 /// See [`ApiError`]
-pub async fn deploy(branch: Option<&str>) -> Result<(), ApiError> {
-    let project = Project::get();
+pub async fn deploy(
+    graph_ref: Option<(String, String, Option<String>)>,
+    branch: Option<String>,
+) -> Result<cynic::Id, ApiError> {
+    let project_definition = match graph_ref {
+        Some((account_slug, project_slug, branch)) => ProjectDefinition::BySlugs {
+            account_slug,
+            project_slug,
+            branch,
+        },
+        None => {
+            let project = Project::get();
+            let project_metadata_file_path = project.dot_grafbase_directory_path.join(PROJECT_METADATA_FILE);
+
+            match project_metadata_file_path.try_exists() {
+                Ok(true) => {}
+                Ok(false) => return Err(ApiError::UnlinkedProject),
+                Err(error) => return Err(ApiError::ReadProjectMetadataFile(error)),
+            }
+
+            let project_metadata_file = read_to_string(project_metadata_file_path)
+                .await
+                .map_err(ApiError::ReadProjectMetadataFile)?;
+
+            let project_metadata: ProjectMetadata =
+                serde_json::from_str(&project_metadata_file).map_err(|_| ApiError::CorruptProjectMetadataFile)?;
 
-    let project_metadata_file_path = project.dot_grafbase_directory_path.join(PROJECT_METADATA_FILE);
+            ProjectDefinition::ById {
+                project_id: project_metadata.project_id,
+                branch,
+            }
+        }
+    };
 
-    match project_metadata_file_path.try_exists() {
-        Ok(true) => {}
-        Ok(false) => return Err(ApiError::UnlinkedProject),
-        Err(error) => return Err(ApiError::ReadProjectMetadataFile(error)),
+    let tar_file = create_tar_file().await?;
+
+    #[allow(clippy::cast_possible_truncation)] // must fit or will be over allowed limit
+    let archive_file_size = tar_file.metadata().await.map_err(ApiError::ReadArchiveMetadata)?.len() as i32;
```

Comment:
Probably not important but should this be a `i32::try_from().unwrap()` rather than an as?  A truncation seems like it might be a lot more subtle (although it's very unlikely so 🤷)

Reply:
Yeah maybe, I just didn't really read how this part was done, copied it out. Maybe I should expect and crash if it's too large?

Reply:
> Maybe I should expect and crash if it's too large?

Yeah, that seems more correct.  Although now we're speaking about it I also wonder if we should have a limit far below i32::MAX.  If someone's trying to upload gigabytes of stuff then something must have gone wrong...